api/controllers/tag.go

```diff
@@ -0,0 +1,201 @@
+package controllers
+
+import (
+	"encoding/json"
+	"net/url"
+	"strconv"
+
+	"github.com/quintilesims/layer0/api/tag"
+	"github.com/quintilesims/layer0/client"
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+	glob "github.com/ryanuber/go-glob"
+	"github.com/zpatrick/fireball"
+)
+
+type TagController struct {
+	TagStore tag.Store
+}
+
+func NewTagController(t tag.Store) *TagController {
+	return &TagController{
+		TagStore: t,
+	}
+}
+
+func (t *TagController) Routes() []*fireball.Route {
+	return []*fireball.Route{
+		{
+			Path: "/tag",
+			Handlers: fireball.Handlers{
+				"GET":    t.ListTags,
+				"POST":   t.CreateTag,
+				"DELETE": t.DeleteTag,
+			},
+		},
+	}
+}
+
+func (t *TagController) DeleteTag(c *fireball.Context) (fireball.Response, error) {
+	var tag models.Tag
+	if err := json.NewDecoder(c.Request.Body).Decode(&tag); err != nil {
+		return nil, errors.New(errors.InvalidRequest, err)
+	}
+
+	if err := t.TagStore.Delete(tag.EntityType, tag.EntityID, tag.Key); err != nil {
+		return nil, err
+	}
+
+	return fireball.NewJSONResponse(200, nil)
+}
+
+func (t *TagController) CreateTag(c *fireball.Context) (fireball.Response, error) {
+	var tag models.Tag
+	if err := json.NewDecoder(c.Request.Body).Decode(&tag); err != nil {
+		return nil, errors.New(errors.InvalidRequest, err)
+	}
+
+	if err := t.TagStore.Insert(tag); err != nil {
+		return nil, err
+	}
+
+	return fireball.NewJSONResponse(200, tag)
+}
+
+func (t *TagController) ListTags(c *fireball.Context) (fireball.Response, error) {
+	query := c.Request.URL.Query()
+
+	tags, err := t.selectTagsFromQuery(query)
+	if err != nil {
+		return nil, err
+	}
+
+	entityTags := tags.GroupByID()
+	if environmentID := query.Get(client.TagQueryParamEnvironmentID); environmentID != "" {
+		filterTagsByEnvironmentID(entityTags, environmentID)
+	}
+
+	if name := query.Get(client.TagQueryParamName); name != "" {
+		filterTagsByName(entityTags, name)
+	}
+
+	if fuzz := query.Get(client.TagQueryParamFuzz); fuzz != "" {
+		filterTagsByIDOrNameGlob(entityTags, fuzz)
+	}
+
+	if version := query.Get(client.TagQueryParamVersion); version != "" {
+		if version == "latest" {
+			filterTagsByLatestVersion(entityTags)
+		} else {
+			filterTagsByVersion(entityTags, version)
+		}
+	}
+
+	filteredTags := models.Tags{}
+	for _, tags := range entityTags {
+		filteredTags = append(filteredTags, tags...)
+	}
+
+	return fireball.NewJSONResponse(200, filteredTags)
+}
+
+func (t *TagController) selectTagsFromQuery(query url.Values) (models.Tags, error) {
+	entityID := query.Get(client.TagQueryParamID)
+	entityType := query.Get(client.TagQueryParamType)
+
+	if entityType != "" && entityID != "" {
+		return t.TagStore.SelectByTypeAndID(entityType, entityID)
+	}
+
+	if entityType != "" {
+		return t.TagStore.SelectByType(entityType)
```

Comment:
My ape-like brain has some difficulty following the flow here. We're calling `selectTagsFromQuery`() from within `ListTags()`, but our return is coming from a different package called from the `selectTagsFromQuery()` helper. This seems inconsistent with how we do things elsewhere within the `controller` package. I've found it a lot more straightforward to see something like:

```
func (j *JobController) DeleteJob(c *fireball.Context) (fireball.Response, error) {
	id := c.PathVariables["id"]
	if err := j.JobStore.Delete(id); err != nil {
		return nil, err
	}

	return fireball.NewJSONResponse(200, nil)
}
```

I know the above example is a bit unfair since it's doing a lot less, but it seems like using the `provider` abstraction here might give us some benefit in clarity. Then we could do all of the `client` interaction in there and keep this abstraction a little more focused. Thoughts?

Reply:
The tag controller is a much different beast than the entity controllers. Even further, the ListTags function is a huge outlier to anything else in the controller package. Moving this logic elsewhere would just be another example of "kicking the complexity down the road" (e.g. the old `logic` package), which we know usually does more harm than good.

Reply:
I get what you're saying, and the more I think about it, the less it would make sense to move this logic to a different package given how this all currently works. 

Anyway, after sitting with this PR for a good chunk of the morning, I get the logic, but I think my comments have more to do with the flow of the functions --- it seems more in the mould of our pre-refactor efforts. Maybe that's acceptable. But if one of the rest of us wanted to add `UpdateTags()` ourselves, I'm guessing it would be pretty painful.